syntax = "proto3";

package raftpb;

// 1. 日志条目 (Log Entry)
message LogEntry {
    uint64 term  = 1; // 产生该日志时的任期
    uint64 index = 2; // 日志在全局历史中的索引
    bytes data   = 3; // 具体命令 (例如 "SET x 10")
}

// 2. RequestVote RPC (拉票)
message RequestVoteRequest {
    uint64 term           = 1;
    int32  candidate_id   = 2; // 这里建议用 int32 对应你的 Node ID
    uint64 last_log_index = 3;
    uint64 last_log_term  = 4;
}

message RequestVoteResponse {
    uint64 term         = 1;
    bool   vote_granted = 2;
}

// 3. AppendEntries RPC (心跳 + 日志复制)
message AppendEntriesRequest {
    uint64 term          = 1;
    int32  leader_id     = 2; // 建议 int32
    
    uint64 prev_log_index = 3; // 紧邻新日志之前的索引
    uint64 prev_log_term  = 4; // 紧邻新日志之前的任期
    
    repeated LogEntry entries = 5; // 真正运送的日志！
    
    uint64 leader_commit = 6; // Leader 的提交进度
}

message AppendEntriesResponse {
    uint64 term    = 1;
    bool   success = 2;
    
    // 冲突优化的字段 (以后做快速恢复用，现在先留着)
    uint64 conflict_index = 3; 
    uint64 conflict_term  = 4;
}

// ==========================================
// 【新增】4. 客户端交互消息 (Client Interaction)
// ==========================================

// 客户端发来的写请求
message ClientRequest {
    bytes data = 1; // 具体的命令，如 "User=Tom"
}

// 服务端返回给客户端的结果
message ClientResponse {
    bool success = 1;       // true 表示 Leader 成功接收并开始处理
    string leader_hint = 2; // 如果 success 为 false，告诉客户端谁是 Leader (比如 "127.0.0.1:5003")
}

// ==========================================
// gRPC 服务接口
// ==========================================

service RaftService {
    // 内部节点交互
    rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);
    rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
    
    // 【新增】外部客户端接口
    // 只有 Leader 能处理这个请求，Follower 应该返回 false
    rpc Propose(ClientRequest) returns (ClientResponse);
}